### PR TITLE
internal: introduce in-place indenting API

### LINKED ISSUE #9903: Switch to in-place indenting
the 

https://github.com/rust-analyzer/rust-analyzer/blob/182a2b8e72eee2faaef0ef17330937bd0edaa4b6/crates/syntax/src/ast/edit.rs#L185-L203

api exist to change indent level on syntax nodes. It works by returning a copy of the node with indentation applied. 

Long after the introduction of that API, we've switched to mutable API for editing syntax nodes, so now the preference is to modify data in-place. We should introduce new trait for in-place indentation changes, and then update all the usages of the old trait, one by one (no need to do everything in one go).

### PR DESCRIPTION
**Pull request by iDawer**
Introduce `edit_in_place::Indent` that uses mutable tree API and intended to replace `edit::AstNodeEdit`.

Closes #9903

### CODE REVIEW

File: crates/ide_assists/src/handlers/convert_iter_for_each_to_for.rs [hunk 1]
```diff
@@ -46,24 +46,25 @@ pub(crate) fn convert_iter_for_each_to_for(acc: &mut Assists, ctx: &AssistContex
     let body = closure.body()?;
 
     let stmt = method.syntax().parent().and_then(ast::ExprStmt::cast);
-    let syntax = stmt.as_ref().map_or(method.syntax(), |stmt| stmt.syntax());
```
Comment (matklad) [inline crates/ide_assists/src/handlers/convert_iter_for_each_to_for.rs:49]: tbh, not a fan of or_else here -- `.syntax()` is an inline field access, it is transparent to the optimizer

File: crates/syntax/src/ast/edit_in_place.rs [hunk 2]
```diff
@@ -487,6 +483,26 @@ fn normalize_ws_between_braces(node: &SyntaxNode) -> Option<()> {
     Some(())
 }
 
+pub trait Indent: AstNode + Clone + Sized {
+    fn indent_level(&self) -> IndentLevel {
+        IndentLevel::from_node(self.syntax())
+    }
+    fn indent(&self, level: IndentLevel) -> &Self {
```
Comment (matklad) [inline crates/syntax/src/ast/edit_in_place.rs:490]:
```suggestion
    fn indent(&self, level: IndentLevel) {
```

guess we shouldn' return anything here, to make this clear that this is a modification.

Comment (iDawer) [reply crates/syntax/src/ast/edit_in_place.rs:490]: Lots of `edit::AstNodeEdit` usages are chained, I thought to support that. But right, it obscures API change.

**Review by matklad — COMMENTED** [review]
bors r+

**Comment by matklad** [comment]
bors retry

**Pull request merged**